### ISSUE TITLE
Feature: Support vscode-languagedetection on the wasm backend

### ISSUE THREAD

**Issue by TylerLeonhardt** [tensorflow/tfjs#5692]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow.js): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS Big Sur
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow.js installed from (npm or script link): npm
- TensorFlow.js version (use command below): 3.9.0 
- Browser version: node 14.17.5
- Tensorflow.js Converter Version: 3.9.0

**Describe the current behavior**

```
Error: Kernel 'StringSplit' not registered for backend 'wasm'
      at Engine.runKernel (node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:4466:19)
      at stringSplit_ (node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:24073:25)
      at Object.stringSplit__op [as stringSplit] (node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:5406:29)
      at executeOp$h (node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:6506:35)
      at /Users/tyleonha/Code/Microsoft/vscode-languagedetection/node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:6642:56
      at /Users/tyleonha/Code/Microsoft/vscode-languagedetection/node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:4394:22
      at Engine.scopedRun (node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:4404:23)
      at Engine.tidy (node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:4393:21)
      at Object.tidy (node_modules/@tensorflow/tfjs-core/dist/tf-core.node.js:10072:19)
      at /Users/tyleonha/Code/Microsoft/vscode-languagedetection/node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:6642:30
      at executeOp$j (node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:6660:7)
      at _loop_1 (node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:7418:31)
      at GraphExecutor.processStack (node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:7444:13)
      at GraphExecutor.<anonymous> (node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:7370:41)
      at step (node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:81:23)
      at Object.next (node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:62:53)
      at fulfilled (node_modules/@tensorflow/tfjs-converter/dist/tf-converter.node.js:52:58)
```

**Describe the expected behavior**

Not throw error.

**Standalone code to reproduce the issue**

So I tried switching my library to using the wasm backend by changing this line:
https://github.com/microsoft/vscode-languagedetection/blob/main/lib/index.ts#L163
(and importing the right package of course)

but it doesn't seem like my model is playing nice with the wasm backend.

Steps are pretty simple:
* clone vscode-languagedetection
* `yarn add -D @tensorflow/tfjs-backend-wasm`
* change the line above and also the import to use `wasm` instead of `cpu`
* run `npm test` in the repo

Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/CodePen/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by mattsoulanille** [comment]
Hi @TylerLeonhardt. You're seeing this error because the `StringSplit` kernel is not implemented in the wasm backend yet. You can see a list of what ops each backend supports in the [TFJS Ops Matrix](https://docs.google.com/spreadsheets/d/1D25XtWaBrmUEErbGQB0QmNhH-xtwHo9LDl59w0TbxrI/edit#gid=0) or each backend's `register_all_kernels.ts` file ([here's the one for wasm](https://github.com/mattsoulanille/tfjs/blob/fix_ios_32_bit/tfjs-backend-wasm/src/register_all_kernels.ts#L107)).

I'll change this issue to a feature request. It's possible there are other ops in your model that are also not yet supported by wasm. If you add them here, we can better track and implement them.

**Comment by TylerLeonhardt** [comment]
@mattsoulanille unfortunately I don't own the model, it's @yoeo over in https://github.com/yoeo/guesslang/tree/master/guesslang/data/model
and I convert it into a tfjs model.

maybe @yoeo could comment on that.

**Comment by mattsoulanille** [comment]
I took a look at the [converted model.json](https://github.com/microsoft/vscode-languagedetection/blob/main/model/model.json), and it looks like the following ops are used. Ones with a checkmark are supported by the wasm backend, and the ones with a question mark did not appear in the ops matrix (I think they might correspond to another op) :

* ? AddV2 (Add and AddN are supported)
* ? BiasAdd
* [x] Cast
* ? ConcatV2 (Concat is supported)
* [x] ExpandDims
* [x] GatherNd
* [x] GatherV2
* [x] GreaterEqual
* [x] NotEqual
* [x] Pack
* [x] Prod
* [x] Reshape
* [x] Select
* ? Shape
* [x] Slice
* [x] Softmax
* [ ] SparseFillEmptyRows
* [ ] SparseReshape
* [ ] SparseSegmentMean
* [ ] SparseSegmentSum
* ? StatelessWhile
* [x] StridedSlice
* [ ] StringToHashBucketFast
* ? TensorListFromTensor
* ? TensorListReserve
* ? TensorListStack
* [x] Tile
* ? Where
* [x] ZerosLike
* [x] _FusedMatMul

It looks like sparse ops will also need to be implemented, along with StringToHashBucketFast.

**Comment by pyu10055** [comment]
THank you Matt for generating this list, all ops should be supported.
ConcatV2 is same as Concat, AddV2, BiasAdd are same as Add.
We might need to register Sparse and string ops, which are CPU ops, they can be supported in wasm by forwarding to CPU impl @ahmedsabie .
The rest or control flow ops that only exist in the model executors.

**Comment by vladmandic** [comment]
Just FYI, I started doing model validation for all models I'm loading  
it compares ops used by model to ops registered by currently selected backend: <https://github.com/vladmandic/human/blob/980285d36367d3ae903455a4884ff5d8ff9d0ea1/src/models.ts#L90>

**Comment by mattsoulanille** [comment]
@vladmandic, that validation script looks very useful. @pyu10055, it sounds like we could be doing this sort of validation in tfjs to give better error messages about missing ops (i.e. a complete list of missing ops instead of just the first one encountered). To avoid introducing a breaking change, we could run the validation the first time the model is run on a new backend instead of when it's loaded. This avoids the case where a user loads a model that doesn't work on the current backend but then switches to a backend that supports the model. Alternatively, or in addition, we could make it a warning instead of an error just in case the model is somehow able to run anyway (although this should never be the case).

**Comment by vladmandic** [comment]
> Alternatively, or in addition, we could make it a warning instead of an error just in case the model is somehow able to run anyway (although this should never be the case).

@mattsoulanille actually, this does happen sometimes - model may have an execution path that simply does not get triggered.
for example, one model I'm using frequently has a missing op `floormod`, but never had an execution error.

**Comment by pyu10055** [comment]
@mattsoulanille @vladmandic backend compatibility validation would be great feature, but it might need to be a runtime validation, as some of the control flow functions or others nodes might not be executed at runtime.  I think it might be good feature for our in-development debugger app @jinjingforever

**Comment by rthadur** [comment]
`StringSplit`  has been added here https://github.com/tensorflow/tfjs/pull/5715 , please check the latest version.
